src/snowflake/cli/api/project/schemas/project_definition.py

```diff
@@ -102,23 +104,53 @@ def _convert_env(
         return ProjectEnvironment(default_env=(env or {}), override_env={})
 
 
-class ProjectDefinition(_DefinitionV11):
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
-        self._validate(kwargs)
+class DefinitionV20(_ProjectDefinitionBase):
+    entities: Dict = Field(
+        title="Entity definitions.",
+    )
+
+    env: Union[Dict[str, str], ProjectEnvironment, None] = Field(
+        title="Environment specification for this project.",
+        default=None,
+        validation_alias="env",
+        union_mode="smart",
+    )
+
+    @field_validator("env")
+    @classmethod
+    def _convert_env(
+        cls, env: Union[Dict, ProjectEnvironment, None]
+    ) -> ProjectEnvironment:
+        if isinstance(env, ProjectEnvironment):
+            return env
+        return ProjectEnvironment(default_env=(env or {}), override_env={})
+
+    @field_validator("entities")
+    @classmethod
+    def validate_entities(cls, entities: Dict) -> Dict:
+        # TODO Add entities validation logic
+        return entities
+
+
+def build_project_definition(**data):
+    """
+    Returns a ProjectDefinition instance with a version matching the provided definition_version value
+    """
+    if not isinstance(data, dict):
+        return
+    version = data.get("definition_version")
+    version_model = get_version_map().get(str(version))
+    if not version or not version_model:
+        # Raises a SchemaValidationError
+        _ProjectDefinitionBase(**data)
+    return version_model(**data)
+
 
-    @staticmethod
-    def _validate(data: Any):
-        if not isinstance(data, dict):
-            return
-        if version := str(data.get("definition_version")):
-            version_model = _version_map.get(version)
-            if not version_model:
-                raise ValueError(
-                    f"Unknown schema version: {version}. Supported version: {_supported_version}"
-                )
-            version_model(**data)
+ProjectDefinition = Union[DefinitionV10, DefinitionV11, DefinitionV20]
```

Comment:
How would the rest of the code use this? Will we be adding a translation layer?

Reply:
It's only used for type checking. If we have a function with this signature:
```python
def project_definition() -> ProjectDefinition
```
The type checker will understand the types it can hold:
```python
definition = get_definition()
definition.definition_version # str | int
definition.native_app # NativeApp | None
```
Up until now when we used `ProjectDefinition` as a type it implied `DefinitionV11`, which was wrong (it could be `DefinitionV10`), but now that we have `DefinitionV20` I wanted to make it clear that it could be any version.
If some code path expects a specific version, it should narrow the type to that version.

Reply:
let's take native_app as example. Are we expecting to see a lot of this in the business logic?
```
if definition.native_app:
   native_app = definition.native_app
else:
  native_app = translate(definition.entity.something['native_app']) # just a random structure
```
Or will it automatically be handled by a translation layer?

Reply:
To be honest, I think it would be clearer if we have DefinitionV1 and DefinitionV2. DefinitionV1 can still have the same inheritance structure as now (DefinitionV1 -> DefinitionV11 -> DefinitionV10 -> Base) and (DefinitionV2 -> DefinitionV20 -> Base).

The rest of the code can then use DefinitionV1 throughout. ProjectDefinition by itself as a combination of V1 and V2 is just confusing. We can use it in places that handle both (Like render_definition) but the rest of the code should be using DefinitionV1 exclusively.

Reply:
We will handle it in future PRs, but in general the business logic layer shouldn't check for definition versions. We will have a version translation layer, so the underlying code will work as is (v1) until we refactor it to use v2 and flip the translation direction. For simple commands that don't require much refactoring we can move to v2->v1 translation directly.

We can discuss the migration plan f2f, but in any case the version narrowing will be done in a single place, and the logic will be guaranteed to receive a specific definition version.

The change in this PR doesn't really affect the existing code in terms of type checking:
```python
@with_project_definition("native_app")
def app_bundle():
    manager = NativeAppManager(
        # project_definition.native_app is guaranteed to be NativeApp
        # because of @with_project_definition check at runtime
        project_definition=cli_context.project_definition.native_app
    )
    ...
```

Reply:
After discussing f2f, I created these types:
```python
ProjectDefinitionV1 = Union[DefinitionV10, DefinitionV11]
ProjectDefinitionV2 = DefinitionV20
ProjectDefinition = Union[ProjectDefinitionV1, ProjectDefinitionV2]
```
And changed `DefinitionManager` to explicitly return `ProjectDefinitionV1`.